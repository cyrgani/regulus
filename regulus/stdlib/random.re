import(math),
import(lists),

global(__stl_rng_state),
=(__stl_rng_state, __builtin_now_nanos_part()),

# returns a random integer
def(rand, _(
    # source: https://en.wikipedia.org/wiki/Xorshift#Example_implementation
    =(x, __stl_rng_state),
    =(x, ^(x, <<(x, 13))),
    =(x, ^(x, >>(x, 7))),
    =(x, ^(x, <<(x, 17))),
    =(__stl_rng_state, x),
    x
)),

# returns a random integer in low..high
# raises if low >= high
def(randrange, low, high, _(
    =(diff, -(high, low)),
    if(<=(diff, 0), error("Range", "called randrange with an empty range")),
    +(low, abs(%(rand(), diff)))
)),

# returns a random element of the given sequence
def(choose, seq, _(
    =(idx, randrange(0, len(seq))),
    index(seq, idx)
)),

# Seeds the RNG with the given value.
# It is not required to seed the RNG before using it, as it automatically uses the current time in nanoseconds as a start.
def(seed, val, _(
    =(__stl_rng_state, val)
)),

# returns a shuffled version of the given sequence
def(shuffle, seq, _(
    for_in(..(0, len(seq)), i, _(
        =(seq, swap(seq, i, randrange(i, len(seq)))),
    )),
    seq
))
