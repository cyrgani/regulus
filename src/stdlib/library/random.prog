_(
    def(random_u16, _(
        # source: https://en.wikipedia.org/wiki/Linear_congruential_generator#Parameters_in_common_use
        =(a, 75),
        =(c, 74),
        =(m, 65537),
        =(X, now_nanos_part()),
        %(+(*(a, X), c), m)
    ))
)