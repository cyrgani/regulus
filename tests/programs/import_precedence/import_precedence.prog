_(
    import("random"),
    not_random(),
    print("expected error:", catch(random_u16())),
)