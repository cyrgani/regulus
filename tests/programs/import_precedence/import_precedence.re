_(
    import(random),
    not_random(),
    print("expected error:", run_or_string_exception(choose(list(0)))),
)