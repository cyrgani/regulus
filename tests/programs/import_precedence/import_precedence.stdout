not random 
expected error: Name: No function `random_u16` found! 
